### ISSUE TITLE
Praat TextGrid output is invalid with boundaries (times are not monotonic)

### ISSUE THREAD

**Issue by psibre** [marytts/marytts#570]
Currently the **TEXT**
```
Foo, bar.

Baz!
```
processed to **PRAAT_TEXTGRID** causes errors when loading them with Praat:
```
$ cat > test.praat << EOF
Read from file... output_10a88ed.TextGrid
printline OK
Read from file... output_90d18d9.TextGrid
> EOF
$ praat --run test.praat 
OK
Error: Wrong xmin 0.79499998 and xmax 0.45499998.
IntervalTier not read.
TextGrid not read.
Data not read from text file “output_90d18d9.TextGrid”.
Script line 1 not performed or completed:
« Read from file... output_90d18d9.TextGrid »
Script “test.praat” not completed.
Praat: script command <<test.praat>> not completed.
```
[output_10a88ed.TextGrid.txt](https://github.com/marytts/marytts/files/371252/output_10a88ed.TextGrid.txt)
[output_90d18d9.TextGrid.txt](https://github.com/marytts/marytts/files/371253/output_90d18d9.TextGrid.txt)

**Comment by psibre** [comment]
This is a regression introduced in 8ef8b6a5ca347a7187fe89efd8baaf56eb69c0e7, so I'm comparing its parent 10a88ed4b495102b7ae87fe71cbdc9ab59b4453c and the current master (90d18d91dd6f0092c7748fefa507bc9e3aa4def4).

**Comment by psibre** [comment]
Here's the **REALISED_ACOUSTPARAMS** diff:
```diff
10c10
< <ph d="255" end="0.39500004" f0="(1,215)(3,210)(5,204)(7,203)(9,202)(11,202)(13,203)(15,205)(17,209)(19,209)(21,208)(23,208)(25,207)(27,206)(29,205)(31,203)(33,202)(35,200)(37,197)(39,195)(41,193)(43,191)(45,189)(47,188)(49,187)(50,186)(52,185)(54,184)(56,184)(58,183)(60,183)(62,183)(64,183)(66,183)(68,183)(70,183)(72,184)(74,184)(76,185)(78,186)(80,188)(82,191)(84,195)(86,198)(88,199)(90,201)(92,202)(94,202)(96,203)(98,202)(100,202)" p="u"/>
---
> <ph d="255" end="0.39499998" f0="(1,215)(3,210)(5,204)(7,203)(9,202)(11,202)(13,203)(15,205)(17,209)(19,209)(21,208)(23,208)(25,207)(27,206)(29,205)(31,203)(33,202)(35,200)(37,197)(39,195)(41,193)(43,191)(45,189)(47,188)(49,187)(50,186)(52,185)(54,184)(56,184)(58,183)(60,183)(62,183)(64,183)(66,183)(68,183)(70,183)(72,184)(74,184)(76,185)(78,186)(80,188)(82,191)(84,195)(86,198)(88,199)(90,201)(92,202)(94,202)(96,203)(98,202)(100,202)" p="u"/>
16c16
< <boundary breakindex="4" duration="125" tone="H-L%"/>
---
> <boundary breakindex="4" duration="400" tone="H-L%"/>
24,26c24,26
< <ph d="60" end="0.58500004" f0="" p="b"/>
< <ph d="155" end="0.74" f0="(3,178)(6,174)(9,170)(12,165)(16,164)(19,162)(22,161)(25,160)(29,159)(32,158)(35,157)(38,156)(41,156)(45,156)(48,155)(51,155)(54,155)(58,154)(61,154)(64,154)(67,153)(70,153)(74,153)(77,153)(80,152)(83,152)(87,152)(90,152)(93,152)(96,151)(100,151)" p="A"/>
< <ph d="165" end="0.90500003" f0="(3,150)(6,150)(9,150)(12,150)(15,150)(18,150)(21,150)(24,150)(27,150)(30,150)(33,150)(36,151)(39,151)(42,151)(45,151)(48,151)(51,151)(54,151)(57,151)(60,151)(63,151)(66,151)(69,151)(72,151)(75,151)(78,151)(81,152)(84,152)(87,152)(90,152)(93,151)(96,151)(100,151)" p="r"/>
---
> <ph d="60" end="0.45499998" f0="" p="b"/>
> <ph d="155" end="0.61" f0="(3,178)(6,174)(9,170)(12,165)(16,164)(19,162)(22,161)(25,160)(29,159)(32,158)(35,157)(38,156)(41,156)(45,156)(48,155)(51,155)(54,155)(58,154)(61,154)(64,154)(67,153)(70,153)(74,153)(77,153)(80,152)(83,152)(87,152)(90,152)(93,152)(96,151)(100,151)" p="A"/>
> <ph d="165" end="0.775" f0="(3,150)(6,150)(9,150)(12,150)(15,150)(18,150)(21,150)(24,150)(27,150)(30,150)(33,150)(36,151)(39,151)(42,151)(45,151)(48,151)(51,151)(54,151)(57,151)(60,151)(63,151)(66,151)(69,151)(72,151)(75,151)(78,151)(81,152)(84,152)(87,152)(90,152)(93,151)(96,151)(100,151)" p="r"/>
32c32
< <boundary breakindex="5" duration="175" tone="L-L%"/>
---
> <boundary breakindex="5" duration="405" tone="L-L%"/>
43,45c43,45
< <ph d="65" end="0.065000005" f0="" p="b"/>
< <ph d="195" end="0.26000002" f0="(2,240)(5,168)(7,164)(10,160)(12,158)(15,158)(17,157)(20,156)(23,155)(25,155)(28,154)(30,154)(33,153)(35,153)(38,152)(41,152)(43,151)(46,150)(48,150)(51,149)(53,148)(56,148)(58,147)(61,146)(64,145)(66,144)(69,144)(71,143)(74,143)(76,143)(79,143)(82,143)(84,143)(87,144)(89,145)(92,145)(94,146)(97,148)(100,150)" p="{"/>
< <ph d="155" end="0.41500002" f0="(20,152)(40,152)(60,151)(80,149)(100,149)" p="z"/>
---
> <ph d="65" end="0.065" f0="" p="b"/>
> <ph d="195" end="0.26" f0="(2,240)(5,168)(7,164)(10,160)(12,158)(15,158)(17,157)(20,156)(23,155)(25,155)(28,154)(30,154)(33,153)(35,153)(38,152)(41,152)(43,151)(46,150)(48,150)(51,149)(53,148)(56,148)(58,147)(61,146)(64,145)(66,144)(69,144)(71,143)(74,143)(76,143)(79,143)(82,143)(84,143)(87,144)(89,145)(92,145)(94,146)(97,148)(100,150)" p="{"/>
> <ph d="155" end="0.415" f0="(20,152)(40,152)(60,151)(80,149)(100,149)" p="z"/>
51c51
< <boundary breakindex="5" duration="45" tone="L-L%"/>
---
> <boundary breakindex="5" duration="405" tone="L-L%"/>
```
[output_10a88ed.xml.txt](https://github.com/marytts/marytts/files/371258/output_10a88ed.xml.txt)
[output_90d18d9.xml.txt](https://github.com/marytts/marytts/files/371259/output_90d18d9.xml.txt)

**Comment by psibre** [comment]
The weird thing is, the issue pops up only because of the `boundary` inserted by the comma between `Foo` and `bar`. But when I run this in an integration test, the boundary is not inserted, so it's tricky to reproduce systematically...

**Comment by psibre** [comment]
So after merging master locally into my fix branch, if I run
```
$ mvn integration-test -pl voice-cmu-slt-hsmm
```
the check *passes* when it shouldn't.

**But** if I run it as a junit test in Eclipse, it fails as expected:
```
java.lang.AssertionError: interval xmax must be bigger than xmin. Expression: (xmax > xmin). Values: xmax = 0.45499998, xmin = 0.79499996
	at org.codehaus.groovy.runtime.InvokerHelper.assertFailed(InvokerHelper.java:404)
	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.assertFailed(ScriptBytecodeAdapter.java:650)
	at marytts.modules.PraatTextGridGeneratorIT$_checkBoundaryTimes_closure1.doCall(PraatTextGridGeneratorIT.groovy:42)
	at sun.reflect.GeneratedMethodAccessor3.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:294)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1019)
	at groovy.lang.Closure.call(Closure.java:426)
	at groovy.lang.Closure.call(Closure.java:442)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.callClosureForLine(DefaultGroovyMethods.java:5236)
	at org.codehaus.groovy.runtime.StringGroovyMethods.eachLine(StringGroovyMethods.java:615)
	at org.codehaus.groovy.runtime.StringGroovyMethods.eachLine(StringGroovyMethods.java:636)
	at org.codehaus.groovy.runtime.StringGroovyMethods.eachLine(StringGroovyMethods.java:595)
	at org.codehaus.groovy.runtime.dgm$1060.invoke(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoMetaMethodSiteNoUnwrapNoCoerce.invoke(PojoMetaMethodSite.java:274)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:56)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at marytts.modules.PraatTextGridGeneratorIT.checkBoundaryTimes(PraatTextGridGeneratorIT.groovy:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
```

**Comment by psibre** [comment]
I think something is screwing up my CLI testing environment...

**Comment by psibre** [comment]
@seblemaguer Let me push the fix branch to upstream so we can both push into a PR.

**Comment by psibre** [comment]
OK, the [CI build](https://travis-ci.org/marytts/marytts/builds/145798369) for #571 fails as expected.

**Comment by psibre** [comment]
Looks like the local .m2/repository is screwing up the test on my local machine, and thanks to caching, on Travis as well (because build [#267](https://travis-ci.org/marytts/marytts/builds/145801238) *should* have passed).

**Comment by psibre** [comment]
So after resetting hard to 579ebb4a93d31eba9c178694532fc2d6227f8df0 and re-running `mvn install`, the test passes as expected. *Then*, after merging master into `fix-570` locally, and re-running `mvn install` *again* (test disabled), the test finally fails as expected on my machine.

[output_579ebb4.TextGrid.txt](https://github.com/marytts/marytts/files/371454/output_579ebb4.TextGrid.txt)
[output_after.TextGrid.txt](https://github.com/marytts/marytts/files/371455/output_after.TextGrid.txt)

**Comment by psibre** [comment]
```diff
--- output_579ebb4.TextGrid.txt	2016-07-19 14:32:02.000000000 +0200
+++ output_after.TextGrid.txt	2016-07-19 14:39:16.000000000 +0200
@@ -1,57 +1,57 @@
 File type = "ooTextFile"
 Object class = "TextGrid"
 xmin = 0.0 
-xmax = 1.5400000499999997 
+xmax = 2.7749999 
 tiers? <exists> 
 size = 1 
 item []: 
 item [1]:
 class = "IntervalTier" 
 name = "phones" 
 xmin = 0.0 
-xmax = 1.5400000499999997 
+xmax = 2.7749999 
 intervals: size = 11 
 intervals [1]:
 xmin = 0.0 
 xmax = 0.14 
 text = "f" 
 intervals [2]:
 xmin = 0.14 
-xmax = 0.39500004 
+xmax = 0.39499998 
 text = "u" 
 intervals [3]:
-xmin = 0.39500004 
-xmax = 0.52000004 
+xmin = 0.39499998 
+xmax = 0.79499998 
 text = "_" 
 intervals [4]:
-xmin = 0.52000004 
-xmax = 0.58500004 
+xmin = 0.79499998 
+xmax = 0.45499998 
 text = "b" 
 intervals [5]:
-xmin = 0.58500004 
-xmax = 0.74 
+xmin = 0.45499998 
+xmax = 0.61 
 text = "A" 
 intervals [6]:
-xmin = 0.74 
-xmax = 0.90500003 
+xmin = 0.61 
+xmax = 0.775 
 text = "r" 
 intervals [7]:
-xmin = 0.90500003 
-xmax = 1.0800000299999999 
+xmin = 0.775 
+xmax = 1.1800000000000002 
 text = "_" 
 intervals [8]:
-xmin = 1.0800000299999999 
-xmax = 1.1450000349999998 
+xmin = 1.1800000000000002 
+xmax = 2.015 
 text = "b" 
 intervals [9]:
-xmin = 1.1450000349999998 
-xmax = 1.34000005 
+xmin = 2.015 
+xmax = 2.215 
 text = "{" 
 intervals [10]:
-xmin = 1.34000005 
-xmax = 1.4950000499999998 
+xmin = 2.215 
+xmax = 2.3699999000000003 
 text = "z" 
 intervals [11]:
-xmin = 1.4950000499999998 
-xmax = 1.5400000499999997 
+xmin = 2.3699999000000003 
+xmax = 2.7749999 
 text = "_" 
```
So the fix for #555 sets the *correct* duration of 400 ms for interval 3 (the `_`), but then the following interval (the `b`) has a *negative* duration.

**Comment by psibre** [comment]
The trouble is really in all the object-oriented Praat TextGrid modeling. The Praat* classes in `marytts.util.data.text` were over-engineered, and designed (somewhat badly) in emulation of Praat's native TextGrid/Tier container model, which suffers from redundancy by tracking absolute start and end times for *each* interval (despite an implicit contiguity assumption), instead of durations (with an optional global offset).

Moreover, the PraatTextGridGenerator and this Praat* code were written [way back before the code was refactored for Maven](../commit/e8cc4243a487e44a123b9855d79ec72e1c6b047b) (and ended up in different modules for some reason), and before there was any proper dependency management. It would have made much more sense to write a separate, independently useful, library to manipulate Praat TextGrids (and perhaps other common formats that are processed in an inconsistent, ad-hoc manner in MaryTTS), and to pull that library in as a dependency.

As it stands, the growing strain between tracking durations on one hand, and absolute end times on the other, throughout the processing pipeline has turned this into a headache that triggered this issue after #555 was fixed.

**Comment by psibre** [comment]
OK, inspecting some unit-selection and HMM voices for German and English before and after the fix (1b5aa116d75a8f61207dd90dd74f3abc7a0eee9b), and after merging with master, leads me to the following observations:
- for HMM voices, the phone durations are a bit off, and the boundary durations are wrong (cf. #555 and, incidentally, #448).
- after the fix, the phone durations are practically the same, and for unit-selection voices, they are imperceptibly shorter (by one datagram?)
- after merging, the phone *and* boundary durations are fully accurate for HMM voices, and for unit selection voices, the slight offset remains, but should not have a noticeable effect.

I'm attaching the inspected files:
[samples.zip](https://github.com/marytts/marytts/files/373026/samples.zip)
